runtime/src/main/cpp/Memory.cpp

```diff
@@ -689,24 +698,38 @@ inline container_size_t objectSize(const ObjHeader* obj) {
   return alignUp(size, kObjectAlignment);
 }
 
+template <typename func>
+inline void traverseObjectFields(ObjHeader* obj, func process) {
+  const TypeInfo* typeInfo = obj->type_info();
+  if (typeInfo != theArrayTypeInfo) {
+    for (int index = 0; index < typeInfo->objOffsetsCount_; index++) {
+      ObjHeader** location = reinterpret_cast<ObjHeader**>(
+          reinterpret_cast<uintptr_t>(obj) + typeInfo->objOffsets_[index]);
+      process(location);
+    }
+  } else {
+    ArrayHeader* array = obj->array();
+    for (int index = 0; index < array->count_; index++) {
+      process(ArrayAddressOfElementAt(array, index));
+    }
+  }
+}
+
+template <typename func>
+inline void traverseReferredObjects(ObjHeader* obj, func process) {
+  traverseObjectFields(obj, [process](ObjHeader** location) {
+    ObjHeader* ref = *location;
+    if (ref != nullptr) process(ref);
+  });
+}
+
 template <typename func>
 inline void traverseContainerObjectFields(ContainerHeader* container, func process) {
   RuntimeAssert(!isAggregatingFrozenContainer(container), "Must not be called on such containers");
   ObjHeader* obj = reinterpret_cast<ObjHeader*>(container + 1);
+
   for (int object = 0; object < container->objectCount(); object++) {
-    const TypeInfo* typeInfo = obj->type_info();
-    if (typeInfo != theArrayTypeInfo) {
-      for (int index = 0; index < typeInfo->objOffsetsCount_; index++) {
-        ObjHeader** location = reinterpret_cast<ObjHeader**>(
-            reinterpret_cast<uintptr_t>(obj) + typeInfo->objOffsets_[index]);
-        process(location);
-      }
-    } else {
-      ArrayHeader* array = obj->array();
-      for (int index = 0; index < array->count_; index++) {
-        process(ArrayAddressOfElementAt(array, index));
-      }
-    }
+    traverseObjectFields(obj, process);
     obj = reinterpret_cast<ObjHeader*>(
       reinterpret_cast<uintptr_t>(obj) + objectSize(obj));
   }
```

Comment:
Now, the only difference between `traverseContainerObjectFields` and `traverseContainerReferredObjects` is that the latter adds instrumenting (null check). 
1. Consider better naming to reflect that.
2. Or revise usage of non-checked version: guess all the call sites add their own checks anyway. Do we really need both?
3. if two policies are definitely needed than make the intention clear, e.g.:

```
template <typename func> auto checked(func process) {
  return [=](auto location) { if (*location) process(*location); };
}
...
traverseObjectFields(obj, checked(process))
```
4. However, a particular instrumenting policy seems to be a property of (or deducted from) the first parameter type, rather than the call site property. For example, can we get rid of null check for array of not nullable elements? Simple overriding may be good enough.

Reply:
Not sure if it is related to this issue.